jitify2.hpp

```diff
@@ -5593,397 +5666,1599 @@ inline bool read_text_file(const std::string& fullpath, std::string* content) {
+  if (flags & ProcessFlags::kAddUsedHeaderWarning) {
+    // Insert a guarded #warning that we can use to see if this header was
+    // actually included during compilation.
+    insert_directive(&tokens, tokens.begin(), "endif");
+    insert_directive(&tokens, tokens.begin(), "warning",
+                     Token(Tt::kIdentifier, "JITIFY_USED_HEADER"),
+                     Token(Tt::kString, "\"" + full_path + "\""));
+    insert_directive(&tokens, tokens.begin(), "ifdef",
+                     Token(Tt::kIdentifier, "JITIFY_USED_HEADER_WARNINGS"));
+  }
+  if (flags & ProcessFlags::kReplacePragmaOnce) {
+    // Note: Must use source itself as unique idenfitier because multiple
+    // filenames may refer to the same file (via copy/symlink/hardlink).
+    replace_pragma_once_with_ifndef(source, &tokens);
+  }
+  if (flags & ProcessFlags::kReplaceStd) {
+    replace_std_with_cuda_std(&tokens);
+  }
+  if (flags & ProcessFlags::kMinify) {
+    // Reconstruct minified source.
+    minify_cuda_source(tokens.begin(), tokens.end(), cxx_standard_year,
+                       processed_source);
+  } else {
+    processed_source->clear();
+    // Reconstruct source.
+    for (const Token& token : tokens) {
+      processed_source->append(token.source_string());
+    }
   }
-  // HACK This is a WAR for some CUB sources including a header they shouldn't.
-  size_t pos = source.find("#include \"../util_device.cuh\"");
-  if (pos != std::string::npos) {
-    source[pos] = '/';  // Comment out the line
-    source[pos + 1] = '/';
+  return {};
+}
+
+}  // namespace parser
+
+namespace detail {
+
+static const char* const kJitifyBuiltinHeaderPrefix = "__jitify_builtin";
+static const char* const kJitifyCallbackHeaderPrefix = "__jitify_callback";
+
+enum class HeaderLoadStatus {
+  kFailed = 0,
+  kAlreadyLoaded = 1,
+  kNewlyLoaded = 2,
+};
+
+// Note: StringMapT is to allow the caller to use StringOrRef instead of
+// std::string in the map.
+template <class StringMapT>
+HeaderLoadStatus load_header(const parser::IncludeName& include,
+                             HeaderCallback header_callback,
+                             const std::vector<std::string>& include_paths,
+                             bool use_builtin_headers, std::string* full_path,
+                             StringMapT* fullpath_to_source) {
+  auto already_loaded = [&](const std::string& fp) {
+    return fullpath_to_source->count(fp);
+  };
+  auto newly_loaded = [&](std::string source) {
+    fullpath_to_source->emplace(*full_path, std::move(source));
+    return HeaderLoadStatus::kNewlyLoaded;
+  };
+  std::string source;
+  // Try loading via callback.
+  *full_path = include.nonlocal_full_path(kJitifyCallbackHeaderPrefix);
+  *full_path = path_simplify(*full_path);
+  if (already_loaded(*full_path)) return HeaderLoadStatus::kAlreadyLoaded;
+  if (header_callback and header_callback(include, &source)) {
```

Comment:
```suggestion
  if (header_callback && header_callback(include, &source)) {
```

*I've been doing our migration with this branch, this line errors in Visual Studio 2022.*